[spark-3.0-snapshot] Basic implementation for CREATE TABLE and ALTER TABLE.

jose-torres opened the pull request:
Basic implementation for CREATE TABLE and ALTER TABLE.

Note that because of some known interactions with Spark, not all the tests quite pass right now. All of the newly added ones to test the DDL functionality pass.

brkyvz commented on build.sbt:
```diff
@@ -22,7 +22,7 @@ crossScalaVersions := Seq("2.12.8", "2.11.12")
 scalaVersion := crossScalaVersions.value.head
 
-resolvers += "Apache Spark snapshot repository" at "https://repository.apache.org/content/repositories/snapshots"
+resolvers += Resolver.mavenLocal
```
you should probably revert this change. The hope is that all fixes will be in Spark 3.0 some time soon

brkyvz commented on src/main/scala/io/delta/DeltaExtensions.scala:
```diff
@@ -0,0 +1,26 @@
+import org.apache.spark.sql.SparkSessionExtensions
+import org.apache.spark.sql.delta.analysis.DeltaAnalysis
+
+class DeltaExtensions extends Function1[SparkSessionExtensions, Unit] {
```
Can you add docs to this? @zsxwing will use this to inject SQL parser commands as well

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/analysis/DeltaAnalysis.scala:
```diff
@@ -0,0 +1,88 @@
+import org.apache.spark.sql.types.{DataType, StructType}
+import org.apache.spark.sql.util.SchemaUtils
+
+// scalastyle:off
```
remove this

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:
```diff
@@ -0,0 +1,343 @@
+
+import scala.util.control.NonFatal
+
+class DeltaCatalog(val spark: SparkSession) extends V2SessionCatalog(spark.sessionState)
```
We should extend the DelegatingCatalogExtension now

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:
```diff
@@ -0,0 +1,343 @@
+      override val properties: util.Map[String, String]) extends StagedTable with SupportsWrite {
+    override def name(): String = ident.name()
+
+    override def abortStagedChanges(): Unit = {}
```
We can have abort clean up the _delta_log directory if we did write out something

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:
```diff
@@ -0,0 +1,343 @@
+
+    override def abortStagedChanges(): Unit = {}
+
+    override def commitStagedChanges(): Unit = {}
```
ideally, this is when the table creation should happen

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaCatalog.scala:
```diff
@@ -0,0 +1,343 @@
+
+  // We keep this as an internal class because it needs to call into some catalog internals for the
+  // eventual table creation.
+  private class StagedDeltaTableV2(
```
I would change the logic in this class a little bit if possible, but we can do that in a follow up.

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala:
```diff
@@ -0,0 +1,102 @@
+
+  override def buildForV1Write(): InsertableRelation = {
+    new InsertableRelation {
+      override def insert(data: DataFrame, overwrite: Boolean): Unit = {
```
overwrite will never be true when using the V1 Write fallback API.

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala:
```diff
@@ -0,0 +1,102 @@
+        }
+        WriteIntoDelta(
+          log,
+          // TODO: other save modes?
```
We only have SaveMode.Overwrite if `truncate()` or `overwrite(filters: Array[Filter])` is called. Otherwise it's always Append mode in DataSourceV2.

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/commands/CreateDeltaTableCommand.scala:
```diff
@@ -0,0 +1,270 @@
+
+
+  private def assertPathEmpty(
+                               sparkSession: SparkSession,
```
nit: indentation

brkyvz commented on src/main/scala/org/apache/spark/sql/delta/sources/DeltaDataSource.scala:
```diff
@@ -43,14 +45,28 @@ class DeltaDataSource
+    DeltaTableV2(DeltaLog.forTable(SparkSession.active, path))
+  }
+
+  override def getTable(
```
this doesn't exist yet, so we should remove it for now

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:
```diff
@@ -0,0 +1,981 @@
+        sql(s"""
+               |CREATE TABLE delta_test(a LONG, b String)
+               |USING delta
+               |OPTIONS('path'='${dir.getCanonicalPath}')""".stripMargin)
```
can we change this to (or also test with) `LOCATION '${dir.getCanonicalPath}'`

brkyvz replied:
I guess there are tests below, so this is fine

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:
```diff
@@ -0,0 +1,981 @@
+
+        def checkDescribe(describe: String): Unit = {
+          // TODO: The order of these columns matches their definition in the DF and not the order
+          // specified in partitionBy, whereas DBR matches the partition by. Is that okay?
```
should be fine

jose-torres commented:
Addressed comments, except as suggested deferring StagedDeltaTableV2 changes to a followup PR. I agree there's probably a better way to structure this class.

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:
```diff
@@ -0,0 +1,981 @@
+
+        // CREATE TABLE without specifying anything works
+        sql(s"CREATE TABLE t1 USING delta LOCATION '${path.getAbsolutePath}'")
+        checkAnswer(spark.table("t1"), Row(1, "a"))
```
this isn't passing tests

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:
```diff
@@ -0,0 +1,981 @@
+      val path = new File(new File(table.storage.locationUri.get), "a=1")
+      assert(path.listFiles().nonEmpty)
+
+      checkDatasetUnorderly(
```
this isn't passing tests.

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala:
```diff
@@ -0,0 +1,486 @@
+
+      val deltaLog = DeltaLog.forTable(spark, TableIdentifier("delta_test"))
+      val snapshot1 = deltaLog.update()
+      assert(snapshot1.metadata.configuration == Map(
```
This isn't passing tests

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala:
```diff
@@ -0,0 +1,486 @@
+
+      val deltaLog = DeltaLog.forTable(spark, TableIdentifier("delta_test"))
+      val snapshot1 = deltaLog.update()
+      assert(snapshot1.metadata.configuration == Map(
```
this isn't passing tests

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala:
```diff
@@ -0,0 +1,486 @@
+    }
+  }
+
+  test("ADD COLUMNS - simple") {
```
this isn't passing tests

brkyvz commented on src/test/scala/org/apache/spark/sql/delta/DeltaAlterTableSuite.scala:
```diff
@@ -0,0 +1,486 @@
+import org.apache.spark.sql.types.{ArrayType, MapType, StructType}
+import org.apache.spark.util.Utils
+
+class DeltaAlterTableSuite extends QueryTest with SharedSparkSession {
```
most of the tests here are not passing. Something seems to be broken

brkyvz replied:
I think it's because you haven't overloaded `loadTable`

brkyvz commented:
Only a 115 failed tests! Not bad. Those will be fixed in OSS hopefully soon. I'm merging this large PR so that we can start adding more stuff.
